phones/tests/models_tests.py

```diff
@@ -315,37 +345,191 @@ def test_create_relaynumber_already_registered_with_service(
+            method="POST",
+            status=412,
+            code=21714,
+        ),
+        None,
+    ]
+
+    RelayNumber.objects.create(user=phone_user, number="+19998887777")
+
+    mock_services.assert_has_calls(
+        [
+            call(twilio_service1_sid),
+            call().phone_numbers.create(phone_number_sid=twilio_incoming_number_sid),
+            call(twilio_service2_sid),
+            call().phone_numbers.create(phone_number_sid=twilio_incoming_number_sid),
+        ]
+    )
+    mock_messages_create.assert_called_once()
+
+    assert django_cache.get("twilio_messaging_service_closed") == twilio_service1_sid
+    assert caplog.messages == ["twilio_messaging_service"]
+    assert caplog.records[0].code == 21714
+
+
+def test_create_relaynumber_skip_to_fallback_service(
+    phone_user, real_phone_us, mock_twilio_client, settings, django_cache, caplog
+):
+    """If a pool has been marked as full, it is skipped."""
+    twilio_incoming_number_sid = "PNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+    twilio_service1_sid = "MGXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX1"
+    twilio_service2_sid = "MGXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX2"
+    settings.TWILIO_MESSAGING_SERVICE_SID = [twilio_service1_sid, twilio_service2_sid]
+    django_cache.set("twilio_messaging_service_closed", twilio_service1_sid)
+    mock_messages_create = mock_twilio_client.messages.create
+    mock_number_create = mock_twilio_client.incoming_phone_numbers.create
+    mock_number_create.return_value = SimpleNamespace(sid=twilio_incoming_number_sid)
+    mock_services = mock_twilio_client.messaging.v1.services
+    mock_messaging_number_create = mock_services.return_value.phone_numbers.create
+
+    RelayNumber.objects.create(user=phone_user, number="+19998887777")
+
+    mock_services.assert_called_once_with(twilio_service2_sid)
+    mock_messaging_number_create.assert_called_once_with(
+        phone_number_sid=twilio_incoming_number_sid
+    )
+    mock_messages_create.assert_called_once()
+    assert django_cache.get("twilio_messaging_service_closed") == twilio_service1_sid
+    assert caplog.messages == []
 
 
-def test_create_relaynumber_canada(phone_user, mocked_twilio_client):
+def test_create_relaynumber_other_messaging_error_raised(
+    phone_user, real_phone_us, mock_twilio_client, test_settings, caplog, django_cache
+):
+    """If adding to a pool raises a different error, it is skipped."""
+    twilio_incoming_number_sid = "PNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+    twilio_service_sid = test_settings.TWILIO_MESSAGING_SERVICE_SID[0]
+    mock_messages_create = mock_twilio_client.messages.create
+    mock_number_create = mock_twilio_client.incoming_phone_numbers.create
+    mock_number_create.return_value = SimpleNamespace(sid=twilio_incoming_number_sid)
+    mock_services = mock_twilio_client.messaging.v1.services
+    mock_messaging_number_create = mock_services.return_value.phone_numbers.create
```

Comment:
suggestion (non-blocking): a lot of this mock setup code seems repeated in each of these tests? refactor into a function?

Reply:
I was able to move the setup of `twilio_incoming_number_sid` to `mock_twilio_client` setup. I was able to remove a lot of setup code, but only saved one line 😆.